Refactor LIFX discovery to make it faster and more reliable

Djelibeybi opened the pull request:
Signed-off-by: Avi Miller <me@dje.li>

## Breaking change

This PR changes the network adapters used to discover LIFX devices.
Instead of enabling all adapters on which LIFX devices appear, this
PR switches to only using the adapter(s) enabled in Home Assistant.




## Proposed change

Currently there is a race condition that occurs when bulbs respond
multiple times to the discovery broadcast and the integration tries 
to register the same bulb that's already mid-registering. 

This PR adds an in-flight discovery throttle that only allows one 
direct connection per bulb at a time to avoid that race condition.

This is faster because there is a limit to how many packets a bulb 
can process per second. So, the fewer packets we send, the faster 
the bulb can process the ones it receives. 

It's more reliable because there are less packets on the network,
less broadcasts over wifi and thus less overall traffic. This allows
us to reduce the amount of message retries as well.

Finally, it eliminates the "does not generate unique ID" error that 
was caused by multiple discoveries trying to register the same entity
at the same time but out of order.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #69168
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Djelibeybi committed:
Refactor LIFX discovery to make it faster and more reliable

This also eliminates the duplicate ID and missing entity ID
errors that are caused by multiple discoveries trying to register
the same entity at the same time.

I've also added myself as a code owner for the integration as
per ADR 0008.

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
I also took the opportunity to reword the various debug log statements so that the connectivity lifecycle is more obvious if you have debug logs enabled. This includes changing the `entity.who` property to use the MAC address instead of the IP address for identification as IP addresses change, but MAC addresses don't.

Djelibeybi commented:
This may resolve #68943 as well.

Djelibeybi commented:
I've also added additional LIFX models to the `homekit` discovery list. The model names were provided by the nice folks over at LIFX who are literally and figuratively keeping the lights on right now.

amelchio commented on homeassistant/components/lifx/light.py:
```diff
@@ -422,16 +426,17 @@ async def register_new_bulb(self, bulb):
                     entity = LIFXWhite(bulb, self.effects_conductor)
 
-                _LOGGER.debug("%s register READY", entity.who)
+                _LOGGER.debug("Connected to %s", entity.who)
                 self.entities[bulb.mac_addr] = entity
+                self.discoveries_inflight.pop(bulb.mac_addr, None)
                 self.async_add_entities([entity], True)
```
I wonder if the race could be fixed by just not adding the entity if it turns out that it was already added while we were getting the product info. Then we would not need the new dictionary.

```suggestion
                if bulb.mac_addr in self.entities:
                    _LOGGER.debug("Ignoring concurrently connected %s", entity.who)
                else:
                    _LOGGER.debug("Connected to %s", entity.who)
                    self.entities[bulb.mac_addr] = entity
                    self.async_add_entities([entity], True)
```

Djelibeybi replied:
I thought about this but I wanted to stop the second or subsequent attempts before they even start to try and reduce overall network traffic. If we only check here, all that discovery is still being done. 

Let me take another look when I'm not on my phone and back at my desk tomorrow.

Djelibeybi replied:
Yeah, this is too late in the process. The idea behind adding `discoveries_inflight` is tha I don't want Home Assistant to send any additional packets to the bulb.

amelchio reviewed (commented):
Thank you for picking up the LIFX integration. Have you considered signing up as code owner?

Djelibeybi commented:
Let me think about that for a bit. Is there anywhere I can read up about what's involved and what the expectations are from the maintainers?

Djelibeybi commented on homeassistant/components/lifx/light.py:
```diff
@@ -608,7 +613,8 @@ async def update_hass(self, now=None):
         self.postponed_update = None
         await self.async_update()
-        self.async_write_ha_state()
+        if self.entity_id is not None:
```
Honestly I have no idea, but it does. I have some theories around possible race conditions that could cause this, but given how much more reliable reconnection is, I don't have to spend the hours trying to work it out. It just self heals already.

bdraco replied:
`self.entity_id` should only ever be `None` if this function is called before the entity is added to hass

Djelibeybi replied:
Well, it is. Logs incoming.

Djelibeybi replied:
I'm going to resolve this conversation as the logs show that adding this check bypasses the error that occurs without it, but I will keep investigating to try and determine why it happens in the first place.

bdraco replied:
This is a bug. Its registering the bulb multiple times because control is returned to the event loop in the else block of `register_bulb` which allows another instance of it to be called before the entity is in `self.entities`:

The check here is stale 
```python
    async def register_bulb(self, bulb):
        """Handle LIFX bulb registration lifecycle."""
        if bulb.mac_addr in self.entities:
```

when the bulb gets discovered again before the code in else block completes

```python3
        else:
            _LOGGER.debug("Connecting to %s (%s)", bulb.ip_addr, bulb.mac_addr)
....
```

Djelibeybi replied:
Which is why I've added a check before we even get here.

bdraco replied:
The above is the cause of the LIFX does not generate unique id errors in your first log.

For this specific issue, the race is the other way around:

The `async_add_entities` creates a task under the hood so if `register_bulb` gets called again, the entity isn't finished adding yet which means it calls the update function prematurely.

Djelibeybi replied:
Both of these are resolved by limiting inflight discovery and it has the additional benefit of reducing network traffic.

Djelibeybi replied:
I want to refactor the entire integration and move the LIFX Manager and discovery out of the light entity so I can support config and diagnostic entities as well, but this at least fixes the current bugs.

bdraco replied:
> Both of these are resolved by limiting inflight discovery and it has the additional benefit of reducing network traffic.

It doesn't seem like it's resolved if we still need this check

Djelibeybi replied:
You're right, this check isn't required any more. I think I fixed this first, then never came back to it after adding the inflight check, which resolved it as well.

Djelibeybi commented:
@amelchio @bdraco here are the logs that show the huge improvement between the current and updated startup reliability and time for my ~60 LIFX bulb install. I think we should merge this so that we improve things immediately and I'll keep working on it to streamline and optimise it too.
[home-assistant-current.log](https://github.com/home-assistant/core/files/8550731/home-assistant-current.log)
[home-assistant-updated.log](https://github.com/home-assistant/core/files/8550732/home-assistant-updated.log)

Djelibeybi commented:
> Have you considered signing up as code owner?

Would that allow me to replace `aiolifx` with something else that may or may not be Photons?

bdraco commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

Djelibeybi commented:
> Please, add your GitHub username to the `manifest.json` of this integration.

Done.

Djelibeybi commented on homeassistant/components/lifx/light.py:
```diff
@@ -378,18 +379,21 @@ async def start_effect(self, entities, service, **kwargs):
-    async def register_new_bulb(self, bulb):
-        """Handle newly detected bulb."""
+        """Allow a single in-flight discovery per bulb."""
+        if bulb.mac_addr not in self.discoveries_inflight:
```
I have a few theories on this as I see the same thing happening with my Photons-powered integration. I _think_ it's because it relies on the zeroconf/HomeKit mDNS to trigger discovery instead of handling it internally, and it may be getting triggered multiple times.

amelchio replied:
My gut feeling is that this is the real issue and this PR is papering over it so it won't get properly fixed.

From your "home-assistant-current.log":
```
2022-04-25 09:45:52 DEBUG (MainThread) [homeassistant.components.lifx.light] 192.168.42.203 register NEW
[...]
2022-04-25 09:45:53 DEBUG (MainThread) [homeassistant.components.lifx.light] 192.168.42.203 register NEW
```

The above should never happen since discoveries are 1 minute apart. Also, the race in `register_new_bulb()` should not be an issue because aiolifx guards us.

All your bulbs register twice so I suspect that you are somehow running two discovery tasks. Unfortunately, I do not see that issue myself and the logging is very sparse so I am unable to figure out what happens :-(

(I can reproduce the "unique IDs" issue by setting an IP alias on my network interface so ifaddr finds two addresses with the same broadcast address)

Maybe this could help to uncover the underlying issue?

```diff
--- a/aiolifx/aiolifx.py
+++ b/aiolifx/aiolifx.py
@@ -1431,6 +1431,9 @@ class LifxScan:
             lifx_discovery = LifxDiscovery(self.loop, manager)
             discoveries.append(lifx_discovery)
             lifx_discovery.start(listen_ip=ip)
+            print(f"Starting discovery on {ip}")
+            import traceback
+            traceback.print_stack()
             tasks.append(self.loop.create_task(manager.lifx_ip()))

         (done, pending) = await aio.wait(tasks, timeout=timeout)
```

Djelibeybi replied:
Yes, I am running multiple discovery tasks because I have both a wired and wifi connection on my development machine. I suspect this happens a lot though, as many Raspberry Pis would be configured the same way.

In my Photons-based integration, I determine the preferred interface for each broadcast network, but `aiolifx` doesn't do this. It just happily starts multiple scans. By limiting the discoveries in-flight, this eliminates the race condition that occurs on misconfigured multi-homed devices.

Djelibeybi replied:
I don't see this as papering over the real issue, though. I can't change how people chose to configure their systems (and it seems I have to take another look at how macOS is configured wrt networking), but this refactor at least nullifies its impact.

amelchio replied:
The setup process with two discoveries is:
```
  wired: discovery request -> 60x discovery response -> feature request -> feature response -> [add entity]
  wifi: discovery request -> 60x discovery response -> feature request -> feature response -> [duplicate ID]
```

This PR proposes to ignore the second discovery response. That saves the unicast feature request for the duplicate entry (and thus avoids the duplicate ID warning):
```
  wired: discovery request -> 60x discovery response -> feature request -> feature response -> [add entity]
  wifi: discovery request -> 60x discovery response -> [ignore]
```

However, what we really want is just a single discovery since that will furthermore save a redundant broadcast and all the matching responses from each bulb:
```
  wired: discovery request -> 60x discovery response -> feature request -> feature response -> [add entity]
  wifi: [idle]
```

So I think the bug is with the network detection (i.e. the ifaddr usage in aiolifx), not really `register_new_bulb()`.

Djelibeybi replied:
Fair enough. I think I can leverage HA's own networking helpers to limit discovery to whichever interface has been selected for mDNS by HA, and then limit aiolifx to that network interface.

amelchio replied:
Awesome. You might find that adjusting the retries does not matter then because bulbs do not get overloaded in the first place.

Djelibeybi replied:
@amelchio four years ago, you added this line:
```
lifx_ip_addresses = await aiolifx().LifxScan(hass.loop).scan()
```

Any objection if I just remove this line and instead, only enable LIFX discovery on network interfaces enabled by Home Assistant?

amelchio replied:
That sounds reasonable. HA did not have any interface selector back then :)

It's a breaking change, of course – but not in a bad way.

Djelibeybi replied:
How much do we want to protect users from themselves? Specifically: if someone choses to manually enable multiple interfaces on the same broadcast network, the current behaviour will re-occur, unless we specifically check for that as well. 

I'm actually happy with this option because I'd document the badness that would occur if such a configuration is chosen by the user and it would mean we could just let Home Assistant's network config be authorative. 

Thoughts?

Djelibeybi replied:
Alternatively, I keep this fix in place as well, which means if someone _does_ enable multiple interfaces that trigger duplicate discovery, we at least ignore some of it at the cost of additional network traffic.

amelchio replied:
For that situation I would propose my one-line fix from yesterday but upgraded to a warning since we now realized that it is a situation to avoid: https://github.com/home-assistant/core/pull/70458#discussion_r857101697

I personally think that's a fine balance between usability and code complexity. However, @bdraco can probably comment on how other integrations handle this (i.e. being enabled for wired/wifi interfaces on the same subnet)

amelchio replied:
I would personally not complicate the code with the in-flight dictionary just to cater for misconfigurations – but I think that is the kind of judgement call that a code owner should be making :-)

Djelibeybi replied:
I see what you did there. :)

bdraco replied:
Checking the configured network interfaces is a nice improvement 👍 

You'll definitely find that you'll still have the same issue since there are still all kinds of "strange" network setups.  We end up with lots of code in zeroconf to handle these type of problems with overlapping responses since even with the ability to configure it, its still quite a common issue.

As much as it would be nice to to have these type of issues, I think you are stuck with either tracking the discoveries to avoid dupes as you have now, or some type of locking (usually less preferred).   I've tried to remove similar code in zeroconf and always end up having to put it back.

Here is a section from RFC6762 that highlights how different implementations can be https://www.rfc-editor.org/rfc/rfc6762.html#section-6.2
```
   Some Multicast DNS responders treat a physical interface with both
   IPv4 and IPv6 address as a single interface with two addresses.
   Other Multicast DNS responders may treat this case as logically two
   interfaces (one with one or more IPv4 addresses, and the other with
   one or more IPv6 addresses), but responders that operate this way
   MUST NOT put the corresponding automatic NSEC records in replies they
   send (i.e., a negative IPv4 assertion in their IPv6 responses, and a
   negative IPv6 assertion in their IPv4 responses) because this would
   cause incorrect operation in responders on the network that work the
   former way.
```

Djelibeybi replied:
Yeah, it makes quite the difference for performance and stability, too. I'm happy keeping the in-flight check as a backup for those network configs that are on the more insane side.

amelchio replied:
@bdraco Can you clarify a bit ... I understand your experience that a request can have multiple responses but is the change here (i.e. to use the configured network interfaces) sufficient to keep _us_ from sending multiple requests in the first place?

amelchio replied:
I am not asking whether people will shoot their feet but rather if they are able to avoid it.

(Like I said, I'm down with you change; just trying to fully understand the situation)

Djelibeybi replied:
Gotcha. It's certainly avoidable: just leave the network config on Auto, or ensure that if you do have multiple interfaces enabled, they're on different IP networks.

bdraco replied:
It is avoidable except when they actually need to broadcast on multiple interfaces because they have multiple subnets.  We have users that have their IoT subnet, their secure subnet, etc so we should expect they are going to turn on multiple interfaces and we may get broadcast traffic from one reflected back on another.

Djelibeybi replied:
Multiple distinct subnets should be fine as LIFX uses UDP broadcast for discovery, which is rarely reflected, repeated or routed.

Multiple enabled interfaces on the same subnet however, will generate duplicate responses.

This is usually what happens when someone enables both wifi and a wired connection on an RPi, for example.

bdraco replied:
> Multiple distinct subnets should be fine as LIFX uses UDP broadcast for discovery, which is rarely reflected, repeated or routed.

Take a look at the python-zeroconf issue queue history, it paints a slightly different view.

> 
> Multiple enabled interfaces on the same subnet however, will generate duplicate responses.

Yes

> 
> This is usually what happens when someone enables both wifi and a wired connection on an RPi, for example.

If you have wired I would considering enabling wifi as well a bad idea but it happens all the time unfortunately.

bdraco reviewed (commented):
The changes look good but I'm traveling until Wednesday so I can't test them with my LIFX devices

Djelibeybi commented:
> The changes look good but I'm traveling until Wednesday so I can't test them with my LIFX devices

This has been an issue for a while now, so a few more days isn't going to make that much difference. :)

bdraco commented on homeassistant/components/lifx/light.py:
```diff
@@ -408,7 +412,7 @@ async def register_new_bulb(self, bulb):
+                _LOGGER.error("Failed to connect to %s", bulb.ip_addr)
                 bulb.registered = False
             else:
                 bulb.timeout = MESSAGE_TIMEOUT
```
In a future PR we should return above, drop the else, and outdent below

Djelibeybi replied:
I _think_ I get what you're suggesting. Will try it out and see what happens.

Djelibeybi replied:
I'm not entirely sure what the point of this block is any more. Instead of just this minor change, I'd prefer to refactor the whole discovery process as a whole, while moving it out of the light entity as well.

amelchio replied:
I can explain but I am not sure which block you are talking about. Will look at the logs now.

Djelibeybi replied:
@amelchio I've mostly worked it out and it's still somewhat relevant given the current discovery process. I got a single `missing entity_id` error when I first tested @bdraco's suggestion, but I can't seem to repeat that. I'd like to get this merged and tested in the beta though, and will continue to optimise where I can.

bdraco commented:
Ideally we merge this before beta cutoff (sometime before Wed morning) so we can get feedback in beta.

We can always revert if it is a problem.

@Djelibeybi Would you please move the manifest changes into a new PR so in the rare event we do have to revert it, we won't revert them as well.

Djelibeybi commented:
@bdraco Yep, I can do that.

Djelibeybi commented:
Manifest changes spun out into https://github.com/home-assistant/core/pull/70614

bdraco commented:
I ordered a few more LIFX devices and a switch so I'll be able to test on Wednesday when I get back to Texas

bdraco committed:
Merge branch 'dev' into refactor-lifx-discovery

Djelibeybi commented:
@bdraco I've officially asked the LIFX team to update the public protocol with the new switch messages: https://github.com/LIFX/public-protocol/issues/6 and I've been talking to folks at LIFX about this directly too. I also have an alternative integration that uses Photons instead of `aiolifx` that I've been working on for a while now, mostly as a learning exercise to get my head around mapping Home Assistant's opinions with Photon's opinions.

bdraco commented:
I'll check back on this tomorrow as I want to make sure Anders has a chance to look at it before hitting merge.

amelchio commented:
> Would that allow me to replace `aiolifx` with something else that may or may not be Photons?

Yeah. Well, it's almost the other way around: doing a big change like that requires you to sign up as code owner to make sure you are notified of the inevitable fallout.

As a code owner, it's somewhat expected that you track new issues. But you're a volunteer and nobody will hold it against you if you cannot donate enough time to resolve everything.

There is no actual rule but in practise being a code owner also means that you get priority in reviews since people will assume that a) you know what you are doing and b) you will be around to fix up any mistakes.

Djelibeybi committed:
Start discovery on enabled interfaces instead of scanned interfaces

This is a potentially breaking change but does resolve several
duplicate discovery issues caused by the race conditions that occur
when mulitiple discoveries happen on the same logical network.

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Merge branch 'refactor-lifx-discovery' of github.com:Djelibeybi/home-assistant-core into refactor-lifx-discovery

Djelibeybi commented:
@bdraco @amelchio I'd appreciate if you could sanity check the code I borrowed from the `yeelight` integration to detect active interfaces is appropriately used here. It works for me and boy, is discovery even faster with only a single active interface. Thanks for the pointers!

amelchio commented:
I have a couple more comments, just no time to make them right now. But we'll get this into the beta, don't worry about that.

bdraco commented on homeassistant/components/lifx/light.py:
```diff
@@ -13,6 +14,7 @@
 import voluptuous as vol
 
 from homeassistant import util
+from homeassistant.components import network
```
Add `network` as a dep in `manifest.json`

`  "dependencies": ["network"],`

Djelibeybi replied:
I'm fairly sure I did.

amelchio reviewed (approved):
This looks good to me now. Just a few minor comments that you can consider but I am not insisting on any change.

I think you should remove the second paragraph of the Breaking Change note though, it is a bit confusing and contradictory (anything vs. unlikely/most) to me and I don't think it adds much anyway.

Djelibeybi commented on homeassistant/components/lifx/light.py:
```diff
@@ -69,8 +71,8 @@
-MESSAGE_TIMEOUT = 1.0
-MESSAGE_RETRIES = 8
+MESSAGE_TIMEOUT = 1
+MESSAGE_RETRIES = 3
```
My plan is to make these adjustable at the config entry level so folks can tweak the knobs to suit their own needs.

Djelibeybi commented on homeassistant/components/lifx/light.py:
```diff
@@ -551,8 +561,8 @@ def name(self):
-        """Return a string identifying the bulb."""
-        return f"{self.bulb.ip_addr} ({self.name})"
+        """Return a string identifying the bulb by name and mac."""
+        return f"{self.name} ({self.bulb.mac_addr})"
```
I really want to switch to using the LIFX serial instead of the MAC address, because that matches the official app.

Djelibeybi commented on homeassistant/components/lifx/light.py:
```diff
@@ -69,8 +71,8 @@
 DISCOVERY_INTERVAL = 60
-MESSAGE_TIMEOUT = 1.0
-MESSAGE_RETRIES = 8
+MESSAGE_TIMEOUT = 1
```
This will be obvious when the knobs are user adjustable

amelchio commented:
@bdraco Are we good to go here or did you want to do tests with your bulbs first?

bdraco reviewed (approved):
I'm traveling and will test it when I get back home tomorrow. Let's merge this now so we get feedback in beta.

I'll ping @Djelibeybi tomorrow if I find anything needs to be adjusted.
